[feature] improve workflow--prepare data and calculate rankings with one command

DiamonDiva opened the pull request:
Per Ravi's request.
A couple other small improvements, such as automatically making the output and/or logging directories if they don't already exist and separating concerns in the prepare_data and calculate_rankings files

DiamonDiva committed:
update requirements and definitions

DiamonDiva committed:
separate args and functional concerns of prepare_data

DiamonDiva committed:
separate args and functional concerns of calculate_rankings

DiamonDiva committed:
add ability to prepare data and calculate rankings with one command

DiamonDiva committed:
use black formatter on new file

DiamonDiva commented:
Code is functionally ready to go as far as I can tell, curious for your thoughts @MichaelGruschke!

I'll add/update the documentation though before merging. And I'll put some of the non-relevant changes to their own PRs--the requirements update and the definitions update

DiamonDiva committed:
add/update documentation

DiamonDiva committed:
reformat

Berenito commented on calculate_rankings.py:
```diff
@@ -39,19 +40,40 @@ def main():
+    :date_str: date of calculation
+    :output_path: path to the folder in which to save the output files
+    """
+    os.makedirs(output_path, exist_ok=True)
```
We can do it with `pathlib.Path` without the need to import `os`.
```suggestion
    output_path.mkdir(exist_ok=True, parents=True)
```

DiamonDiva replied:
👍

Berenito commented on calculate_rankings.py:
```diff
@@ -39,19 +40,40 @@ def main():
         divisions = [args.division]
 
-    setup_logger(args.output / f"calculate_rankings-{args.season}-{args.division}-{date_str}.log")
+    calculate_rankings(args.input, args.season, divisions, date_str, args.output)
```
I don't see much advantage in having this as a separate function. When we have the script with `main` function, I would expect to have everything important written there so I don't have to look through the code to find the actual docstring. And you just redefine the parameters.

DiamonDiva replied:
I feel it separates concerns on an important level:
- main should only be run if one explicitly would like to run this script--the calculate rankings script. This takes care of acquiring the necessary values to run this script via the arguments
- the calculate_rankings function now does the actual calculation with the necessary parameters, regardless of where those parameters come from or how they were generated

This becomes much more relevant for something like the new workflow where we can then specifically call the `calculate_rankings` function with the appropriate parameters and not have to worry about unrelated, calculate_rankings-script-specific user-interface concerns. If anything, I think it makes the code more readable because it removes the user-interaction code from the actual logic and leaves a function that is dedicated to calculating rankings

Not a hill I'll die on, but I do feel this is a clearer and cleaner format. Happy to chat about it on a call if you'd like

Berenito replied:
I personally prefer the function to do a single thing, which this approach violates as we are doing parsing, logging, calculating, saving, etc. in a single function. Main is the exception, as it is basically the execution of a script. I like the idea of a single unified parser defined elsewhere, though.

DiamonDiva replied:
I think we agree about ideas here and disagree about methods. I also find the function to be way too busy and doing too many things, and this is simply a first step in pulling out parsing. This pull request is more about adding a feature than about refactoring the code, and so I'm doing the small amount of refactoring that moves us towards single-purpose-functions and aligns with giving Ravi easier use

Berenito commented on calculate_rankings.py:
```diff
@@ -39,19 +40,40 @@ def main():
-            date=args.date,
+            input_path / f"{season}-{division}-euf-games.csv",
+            name=f"{season}-{division}-euf",
+            date=date_str,
```
Here we need to have date with the separators for correct filtering of the games in the dataset, where the format with `-` is expected.

DiamonDiva replied:
I see, so it'd be better to take the date pre-formatted and do the formatting in this function to be safe?

Berenito commented on prepare_data.py:
```diff
@@ -30,54 +26,76 @@ def main():
+    :divisions: list of divisions for which the ranking should be calculated
+    :output_path: path to the folder in which to save the output CSVs
+    """
+    os.makedirs(output_path, exist_ok=True)
```
```suggestion
    output_path.mkdir(exist_ok=True, parents=True)
```

Berenito commented on prep_and_calculate.py:
```diff
@@ -0,0 +1,120 @@
+
+    calc_input_path = prep_output_path = output_path / "data_preparation"
+
+    prepare_data(input_path, season, divisions, prep_output_path)
```
You can call this as `prepare_data.main(args)` and then you change the corresponding main function to `main(args: t.Optional[argparse.Namespace] = None)`

DiamonDiva replied:
this is related to the discussion above about separating out the args parsing and the important procedures in `prepare_data` and `calculate_rankings`. I can go this route if we decide that this separation of concerns doesn't make sense

MichaelGruschke replied:
I would prefer explicit parameters instead of a god object, especially if we were to cover this with tests.

Berenito replied:
Regarding the tests, for now there is a test folder with the testing (manual) of the correctness of the algorithms to their "official" versions.

DiamonDiva replied:
I think we all agree with Michael's statement about preferring no oversized, over-purposed objects and methods (see longer thread above). For this pull request, the question is whether or not parsing should already be extracted. Other improvements can be dealt with later

Berenito reviewed (changes requested):
Thanks for your PR, is good to actually collaborate on something here ;).
However, I don't like much the change when we add another "almost main" function to the runnable script, I would prefer if we keep it as before.
It is still possible to run such script from the `prep_and_calculate` file, you just add optional `args` parameter to `main` in `prepare_data` and `calculate_rankings`.
I see the changes in the other small PRs are also included here, so I just close them and we merge everything here.

MichaelGruschke commented on prepare_data.py:
```diff
@@ -30,54 +26,76 @@ def main():
     for division in divisions:
-        logger.info(f"Preparing data for season {args.season}, {division} division.")
-        df_teams = pd.read_csv(args.input / f"teams-{division}.csv")
+        logger.info(f"Preparing data for season {season}, {division} division.")
```
small nit pick, but imho you should generally take advantage of lazy logging in python
for reference: https://medium.com/flowe-ita/logging-should-be-lazy-bc6ac9816906

Berenito replied:
I personally prefer cleaner and better-looking string formatting to some negligible time improvement.

DiamonDiva replied:
Regardless of whichever route we think is best, this pull request is just changing the line so as not to introduce a bug. How to deal with logging should not be part of this pull request in my mind

MichaelGruschke commented on prep_and_calculate.py:
```diff
@@ -0,0 +1,120 @@
+    )
+
+
+def prep_and_calculate(input_path, season, divisions, date_str, output_path):
```
type hints would be nice

DiamonDiva replied:
true

MichaelGruschke commented on prepare_data.py:
```diff
@@ -30,54 +26,76 @@ def main():
+    * Add to the same folder a TXT file with the pairs <team>, <tournament>; specifying that the given team has met the
+      EUF roster requirements for the particular tournament (filename should be teams_at_tournaments-<division>.txt)
+
+    Procedure:
```
You list all the logical steps performed by the function.
This is a strong indicator, that the the whole code should be split up into exactly these steps as individual functions.
But this is probably not in scope of this PR. Maybe not even worth it.

Berenito replied:
Generally the approach is not wrong, but I would say for ~100 lines of code it is definitely not worth it (maybe if we decided to include some unit tests, it would make sense).

DiamonDiva replied:
this is just copy and paste to keep with the way it was done in the other files. I agree with what you're suggesting, but that's more of a dedicated change/refactoring and the purpose of this PR is to allow Ravi to enter a single command and smoothly run data preparation and ranking calculation

MichaelGruschke commented:
@DiamonDiva 
One idea: How about we have a single point of entry, a main.py file that parses all necessary arguments and remove the main functions from [calculate_rankings.py](https://github.com/Berenito/Ranking/blob/season2024/calculate_rankings.py) and [prepare_data.py](https://github.com/Berenito/Ranking/blob/season2024/prepare_data.py).
In this main.py you can pass an additional parameters specifying whether to run prepare_data or calculate_ranking or both.
This way you don't have to maintain 3 argparsers when changing the cli in the future.

Berenito commented:
> @DiamonDiva One idea: How about we have a single point of entry, a main.py file that parses all necessary arguments and remove the main functions from [calculate_rankings.py](https://github.com/Berenito/Ranking/blob/season2024/calculate_rankings.py) and [prepare_data.py](https://github.com/Berenito/Ranking/blob/season2024/prepare_data.py). In this main.py you can pass an additional parameters specifying whether to run prepare_data or calculate_ranking or both. This way you don't have to maintain 3 argparsers when changing the cli in the future.

You got a point with the almost-the-same argparser used in all the functions, we should definitely define it in some common file, maybe `utils/parsing.py` or something similar.

DiamonDiva commented:
sorry, somehow missed the larger comments down here. I like that, will get to it later today probably!

DiamonDiva committed:
use path instead of os, add types to func definition

DiamonDiva committed:
make date small d because it is required

DiamonDiva committed:
get rid of all args stuff and main() in calculate_rankings

DiamonDiva committed:
get rid of args stuff and main() in prepare_data

DiamonDiva committed:
convert prep_and_calculate to just main() function, and flexibly able to call prep, calc, or both

DiamonDiva committed:
rename prep_and_calculate to main

DiamonDiva committed:
reformatting

DiamonDiva commented:
@MichaelGruschke @Berenito 

I think that I have implemented all the changes that were not contentious:
- not using os anymore
- added types to function definitions
- changed where "date" is reformatted
- got rid of main() functions and any direct calling of `prepare_data` and `calculate_rankings` files
- changed `prep_and_calculate` to `main`, simplified it to only a `main()` function, and added the ability to specify if `prepare_data`, `calculate_rankings`, or both are called, with the default being both

I think this also lands in a reasonable middle-ground for this pull request in terms of reducing the responsibilities of the primary functions in each of `prepare_data` and `calculate_rankings` but not fully refactoring out the responsibilities of those functions

let me know your thoughts and if there's any other changes you'd suggest!

DiamonDiva committed:
remove argparse from files where unused

DiamonDiva commented:
> > @DiamonDiva One idea: How about we have a single point of entry, a main.py file that parses all necessary arguments and remove the main functions from [calculate_rankings.py](https://github.com/Berenito/Ranking/blob/season2024/calculate_rankings.py) and [prepare_data.py](https://github.com/Berenito/Ranking/blob/season2024/prepare_data.py). In this main.py you can pass an additional parameters specifying whether to run prepare_data or calculate_ranking or both. This way you don't have to maintain 3 argparsers when changing the cli in the future.
> 
> You got a point with the almost-the-same argparser used in all the functions, we should definitely define it in some common file, maybe `utils/parsing.py` or something similar.

For now it is simply in the new `main.py` file since it is the only file using `argparse` now. Can be pulled entirely out to a utils file if you'd prefer

Berenito commented on main.py:
```diff
@@ -0,0 +1,96 @@
+        "--methods",
+        "-m",
+        default="both",
+        choices=["prep", "calc", "both"],
```
Please, rather use full words instead of abbreviations.

Berenito reviewed (approved):
OK, looks much better. 
I am fine with the way it is now, just set more reasonable choices.
But I would suggest the following changes in the `main.py` function soon-ish (if both of you do not strongly oppose ;) ):
* Move the code from `calculate_ratings` and `prepare data` to `main.py` (I don't like this idea of saving files inside a function)
* Take out some logical parts from `prepare_data `and define in separate functions outside `main()` in `main.py`

Berenito commented:
> > > @DiamonDiva One idea: How about we have a single point of entry, a main.py file that parses all necessary arguments and remove the main functions from [calculate_rankings.py](https://github.com/Berenito/Ranking/blob/season2024/calculate_rankings.py) and [prepare_data.py](https://github.com/Berenito/Ranking/blob/season2024/prepare_data.py). In this main.py you can pass an additional parameters specifying whether to run prepare_data or calculate_ranking or both. This way you don't have to maintain 3 argparsers when changing the cli in the future.
> > 
> > 
> > You got a point with the almost-the-same argparser used in all the functions, we should definitely define it in some common file, maybe `utils/parsing.py` or something similar.
> 
> For now it is simply in the new `main.py` file since it is the only file using `argparse` now. Can be pulled entirely out to a utils file if you'd prefer

I would say it is ok for now, until we don't have more runnable scripts.

DiamonDiva committed:
use full words

DiamonDiva committed:
move prepare_data and calculate_rankings to main.py

DiamonDiva committed:
pull out some logic to individual functions

DiamonDiva commented:
> OK, looks much better. I am fine with the way it is now, just set more reasonable choices. But I would suggest the following changes in the `main.py` function soon-ish (if both of you do not strongly oppose ;) ):
> 
>     * Move the code from `calculate_ratings` and `prepare data` to `main.py` (I don't like this idea of saving files inside a function)
> 
>     * Take out some logical parts from `prepare_data `and define in separate functions outside `main()` in `main.py`

I'm not sure I fully understand what you meant, but I pulled `calculate_ratings()` and `prepare_data()` over to `main.py`, removed `calculate_ratings.py` and `prepare_data.py`, and then pulled a couple smaller logical pieces out of `prepare_data()` into separate functions.

Does this feel like a good place for now? If so, I'd merge and let Ravi know he can pull the latest and start having an easier workflow, and then we can further refactor as we see fit

Berenito commented:
Yeah, I think we can merge it now, good job.
